### PR TITLE
Gemini String flattening issue fix

### PR DESCRIPTION
**Pull request by sid-sagacollab**
# Description
This is a PR to fix the flat string being passed to the gemini client without specifying roles.

# Testing

Please provide a list of the ways you can "access" or use the functionality. Please try and be exhaustive here, and make sure that you test everything you list.

- [x] I have tested this on real data that is reasonable and large
- [ ] If I changed the interaction with JupyterLab, I tested that it does not break other programs (like VS Code), and tested that it works "multiple times" in the same notebook.

# Documentation

N/A

### CODE REVIEW

File: mito-ai/mito_ai/gemini_client.py [hunk 1]
```diff
@@ -1,14 +1,86 @@
...
+                        text_content = item.get("text", "")
+                        if text_content:
+                            parts.append({"text": text_content})
+                    elif item.get("type") == "image_url":
+                        image_url_data = item.get("image_url", {})
+                        image_url = image_url_data.get("url", "")
+
+                        if image_url:
+                            # Handle base64 data URLs
+                            if image_url.startswith("data:"):
```
Comment (aarondr77) [inline mito-ai/mito_ai/gemini_client.py]:
```suggestion
                        if image_ur and  image_url.startswith("data:"):
```

**Review by aarondr77 — CHANGES_REQUESTED** [review]
I'm getting this error every time I try to use Gemini. I know that you're doing a lot of manual testing, but I think your efforts would be better spent on creating pytests. I think a good method for a PR like this would be to: figure out what the correct response type should be, create pytests for it, then implement the function. 

Doing this test-driven development strategy will probably: 1) Save you a ton of time in development because you can just run your pytests and test all of the edge cases after each change without having to spend time manually testing 2) Make it more easier for Cursor agent to write some of the implementation code for you 3) Save us a lot of time in the review cycles because the first round of reviews won't be about bug catching and we can skip straight to the code review part. 

Ideally, we are only doing 1 review per PR. 

Screenshot 2025-06-10 at 11 49 22 AM

File: mito-ai/mito_ai/gemini_client.py [hunk 2]
```diff
@@ -17,71 +88,80 @@ def __init__(self, api_key: Optional[str], model: str):
...
+                        config=GenerateContentConfig(
+                            system_instruction=""   # No system instruction needed for inline completions
+                        )
+                    )
             else:
-                # Only pass provider_data to the server
                 return await get_gemini_completion_from_mito_server(
                     model=provider_data["model"],
-                    contents=provider_data["contents"],
+                    contents=str(contents),  # Use the processed contents instead of str(messages)
```
Comment (aarondr77) [inline mito-ai/mito_ai/gemini_client.py]: In the openai_client, we just leave this as the list of contents and then then the `get_ai_completion_from_mito_server` function just converts the entire request body like this `body=json.dumps(data), `

Do we need to convert it to a string here because `types.Part.from_bytes` is not JSON serializable or something? If so, can you document that here.

File: mito-ai/mito_ai/gemini_client.py [hunk 3]
```diff
@@ -17,71 +88,80 @@ def __init__(self, api_key: Optional[str], model: str):
...
+                        )
+                    )
+                else:
+                    response = self.client.models.generate_content(
+                        model=provider_data["model"],
+                        contents=contents,
+                        config=GenerateContentConfig(
+                            system_instruction=""   # No system instruction needed for inline completions
+                        )
+                    )
```
Comment (aarondr77) [inline mito-ai/mito_ai/gemini_client.py]: Would this simplify things?

Screenshot 2025-06-10 at 4 08 00 PM

File: mito-ai/mito_ai/tests/test_gemini_client.py [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+import pytest
+from mito_ai.gemini_client import extract_system_instruction_and_contents
+
+# Dummy base64 image (1x1 PNG)
+DUMMY_IMAGE_DATA_URL = (
+    "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAQAAAC1HAwCAAAAC0lEQVR42mP8/wcAAgMBAp9l9AAAAABJRU5ErkJggg=="
+)
+
+def test_mixed_text_and_image():
```
Comment (aarondr77) [inline mito-ai/mito_ai/tests/test_gemini_client.py:12]: yay! Thanks for creating these!

**Review by aarondr77 — APPROVED** [review]
Feel free to implement these small changes if they seem reasonable and then merge it in.
